pkg/controllers/management/authprovisioningv2/cluster.go

```diff
@@ -44,20 +61,107 @@ func (h *handler) OnCluster(key string, cluster *v1.Cluster) (*v1.Cluster, error
 	existingRole, err := h.roleCache.Get(cluster.Namespace, roleName)
 	if err != nil {
 		if !k8serrors.IsNotFound(err) {
-			return nil, err
+			return err
 		}
 
 		if _, err := h.roleController.Create(role); err != nil && !k8serrors.IsAlreadyExists(err) {
-			return nil, err
+			return err
 		}
-		return cluster, nil
+		return nil
 	}
 
 	if !reflect.DeepEqual(existingRole.Rules, role.Rules) {
 		existingRole = existingRole.DeepCopy()
 		existingRole.Rules = role.Rules
 		_, err := h.roleController.Update(existingRole)
-		return nil, err
+		return err
 	}
-	return cluster, nil
+
+	return nil
+}
+
+func (h *handler) createClusterAdminRole(cluster *v1.Cluster) error {
+	roleName := rbac.ProvisioningClusterAdminName(cluster)
+	role := &rbacv1.Role{
+		TypeMeta: metav1.TypeMeta{},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      roleName,
+			Namespace: cluster.Namespace,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: cluster.APIVersion,
+					Kind:       cluster.Kind,
+					Name:       cluster.Name,
+					UID:        cluster.UID,
+				},
+			},
+		},
+		Rules: []rbacv1.PolicyRule{
+			{
+				APIGroups:     []string{cluster.GroupVersionKind().Group},
+				Resources:     []string{"clusters"},
+				ResourceNames: []string{cluster.Name},
+				Verbs:         []string{"*"},
+			},
+		},
+	}
+
+	existingRole, err := h.roleCache.Get(cluster.Namespace, roleName)
+	if err != nil {
+		if !k8serrors.IsNotFound(err) {
+			return err
+		}
+
+		if _, err := h.roleController.Create(role); err != nil && !k8serrors.IsAlreadyExists(err) {
+			return err
+		}
+		return nil
+	}
+
+	if !reflect.DeepEqual(existingRole.Rules, role.Rules) {
+		existingRole = existingRole.DeepCopy()
+		existingRole.Rules = role.Rules
+		_, err := h.roleController.Update(existingRole)
+		return err
+	}
+
+	return nil
+}
+
+func (h *handler) cleanClusterAdminRoleBindings(cluster *v1.Cluster) error {
+	logrus.Infof("cleanClusterAdminRoleBindings %v ", cluster.Name)
+	// Collect all the errors to delete as many rolebindings as possible
+	var allErrors []error
+
+	roleName := rbac.ProvisioningClusterAdminName(cluster)
+	existingRole, err := h.roleCache.Get(cluster.Namespace, roleName)
+	if err != nil {
+		if !k8serrors.IsNotFound(err) {
+			return err
+		}
+	}
+	if existingRole == nil {
+		logrus.Infof("Role with ProvisioningClusterAdminName not found %v ", roleName)
+		return nil
+	}
```

Comment:
I would expect the role to be gone, or at least getting deleted so I don't think this part is necessary. The role being gone does not mean the roleBindings are cleaned up

Reply:
removed this check